experiment/simulation/js/flipflop.js

```diff
@@ -0,0 +1,780 @@
+        }
+        else if (pos === "Q'") {
+            this.qbarIsConnected = val;
+        }
+    }
+
+    generateOutput() {
+        const d = this.dbefore;
+        const clk = getOutputD(this.clk[0], this.clk[1]);
+        if (this.prevclk === true && clk === false) {
+            if (!d) {
+                this.q = false;
+                this.qbar = true;
+            }
+            else if (d) {
+                this.q = true;
+                this.qbar = false;
+            }
+            this.prevclk = false;
+        }
+        else if (this.prevclk === false && clk === true) {
+            this.prevclk = true;
+        }
+        
+    }
+}
+
+
+function addDFlipFlop(x, y) {
+    const ff = new DFlipFlop();
+    ff.registerComponent("working-area", x, y);
+}
+
+window.addDFlipFlop = addDFlipFlop;
+
+export function getOutputD(gate, pos) {
+    if (pos === "Q") {
+        return gate.q;
+    }
+    else if (pos === "Q'") {
+        return gate.qbar;
+    }
+    // But if the gate is not an FA, but an input bit, then return the value of the input
+    else {
+        return gate.output
+    }
+}
+
+// done checking
+export function getResultDD(ff) {
+    // check if flipflop type is Gate object
+    if (ff.constructor.name === "Gate") {
+        return;
+    }
+    if (ff.pr.length > 0) {
+        if (ff.pr[0].id === "Input-0" && ff.pr[0].output === false) {
+            ff.q = true;
+            return;
+        }
+    }
+    if (ff.clr.length>0) {
+        if (ff.clr[0].id === "Input-0" && ff.clr[0].output === false) {
+            ff.q = false;
+            return;
+        }
+    }
+    if (getOutputD(ff.d[0], ff.d[1]) != null && getOutputD(ff.clk[0], ff.clk[1]) != null) {
+            ff.generateOutput();
+    }
+    return;
+}
+
+
+// done checking
+export function checkConnectionsDD() {
+    let correctConnection = true;
+    for (let ffID in flipFlops) {
+        const gate = flipFlops[ffID];
+        // For Full Adder objects
+        // Check if all the outputs are connected
+        const id = document.getElementById(gate.id);
+
+        if(gate.pr.length != 0 && gate.clr.length != 0){
+            correctConnection = false;
+            printErrors("Can't activate both preset and clear\n",id);
+            break;
+        }
+
+        if (gate.qIsConnected == false) {
+            correctConnection = false;
+            printErrors("Q of flip flops must be connected\n",id);
+            break;
+        }
+        // Check if all the inputs are connected
+        if (gate.d == null || gate.d.length == 0) {
+            correctConnection = false;
+            printErrors("D of flip flops must be connected\n",id);
+            break;
+        }
+        if (gate.clk == null || gate.clk.length == 0) {
+            correctConnection = false;
+            printErrors("CLK of flip flops must be connected\n",id);
+            break;
+        }
+    }
+    let count=0;
+    let oriVal=false;
+    for (let gateId in gates) {

```

Comment:
Please ensure that you need to use the for..in loop and not the for..of loop

Reply:
using for..of in those places instead of for..in throws error saying that the list is not iterable, so in such places we have used for..in only